### DISCUSSION TITLE
Raspberry PI :  error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1

### DISCUSSION THREAD

**Discussion by leonardoee** [pymupdf/PyMuPDF#947] [Q&A] [answered]
Hi,
I've been facing the a similar issue from [https://github.com/pymupdf/PyMuPDF/issues/414#issue-532885738]
 when installing PyMuPDF on my raspberry Pi. I've been trying work this out for more than a week. Just to highlight, I'm a beginner on coding and linux.

Firstly when I ran "pip install PyMuPDF" I got the error bellow:
"command **'arm-linux-gnueabihf-gcc'** failed with exit status 1"
The same error happened with I downloaded and tried to install PyMuPDF-master by "python setup.py install"

After several unsuccessful tries and errors installing all sorts of packages when I ran "python setup.py install" from PyMuPDF-master I got a slightly different error:
"error: command **'gcc'** failed with exit status 1"

I did all the steps on the links bellow, but nothing worked.
https://solarianprogrammer.com/2018/05/06/building-gcc-cross-compiler-raspberry-pi/
https://github.com/Pro/raspi-toolchain

Appreciate any help on this.

**Comment by JorjMcKie** [comment]
There are fundamental differences in the install methods you used. When installing PyMuPDF you have two options:
1. install from a (pre-compiled) "wheel"
2. install from sources

# Ad 1.
Wheel refers to the extensions `.whl` of the single installation file required. It is a ZIP format. It can contained any mixture of Python files, binary pre-compiled Python extensions and potentially arbitrary other files.
The installation steps for wheels require **_no compiler_** and basically are just file copies to the respective Python `dist-packages` or `site-packages` folder.
I am preparing such wheels for PyMuPDF. Every combination of target operating system, Python version and bitness needs its own specific wheel file. There are currently 16 wheel files: Linux and OSX 4 each for the Python 64bit versions 3.6, 3.7, 3.8, 3.9, and 8 for Windows - same Python versions, but each with 32bit and 64bit.
If you execute `python -m pip install pymupdf` on any platform, PIP will determine the platform on which it has been invoked, generate a file name from this information and download the matching wheel from PyPI. If that file is not on PyPI, then pip will download the PyMuPDF source installation material from PyPI and execute `python setup.py install` with it. This is equivalent to alternative 2 following.

# Ad 2.
You can generate PyMuPDF from sources. This obviously requires a C compiler **_suitable for the desired target platform._** In addition, you must download and generate MuPDF (Artifex Software Inc.) before generating PyMuPDF.

Obviously, a lot can go wrong with this approach.

You did not specify any details of the encountered errors. So there is no way to tell what really happened. But they all indicate, that the "source generation" alternative has been tried to execute.
And of course that was **_bound to fail_**, if no previous MuPDF installation existed.

So what may be next steps:
* Either install build MuPDF. This must happen with a C compiler suitable for the intended target environment. Maybe other adjustments to the MuPDF build process are required, too.
* Or, build a wheel that can work with your RPI. If this can be done, you are out of trouble.

I am currently developping additional wheels for ARM environments. Maybe there is a chance to make one for you.
What is the bitness of this RPI environment: 32bit / 64bit and the Python version?

**Comment by leonardoee** [comment]
Hi,
Firstly, thank you for this quick turnaround.
This library is great. I was amazed when I saw my script ran in 1/3 of the time compared to when I was using PyPDF2 + reportlab.
I'm running `Raspbian GNU/Linux 10 (buster) - 32bits`, kernel: `Linux raspberrypi 5.10.17-v7l+ #1403` armv7l GNU/Linux and `python 3.7.3`.
I have no issues on changing the python version to match an existent wheel.
Bellow is the error I've been facing when I run `python -m pip install pymupdf`

 pi@raspberrypi:~/dev/guidegf/sggen $ `python3 -m pip install pymupdf`
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting pymupdf
  Using cached PyMuPDF-1.18.9.tar.gz (308 kB)
Building wheels for collected packages: pymupdf
  Building wheel for pymupdf (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/pi/dev/guidegf/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/setup.py'"'"'; __file__='"'"'/tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-4ualunyb
       cwd: /tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/
  Complete output (20 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-armv7l-3.7
  creating build/lib.linux-armv7l-3.7/fitz
  copying fitz/__init__.py -> build/lib.linux-armv7l-3.7/fitz
  copying fitz/fitz.py -> build/lib.linux-armv7l-3.7/fitz
  copying fitz/utils.py -> build/lib.linux-armv7l-3.7/fitz
  copying fitz/__main__.py -> build/lib.linux-armv7l-3.7/fitz
  running build_ext
  building 'fitz._fitz' extension
  creating build/temp.linux-armv7l-3.7
  creating build/temp.linux-armv7l-3.7/fitz
  arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/mupdf -I/usr/local/include/mupdf -Imupdf/thirdparty/freetype/include -I/home/pi/dev/guidegf/include -I/usr/include/python3.7m -c fitz/fitz_wrap.c -o build/temp.linux-armv7l-3.7/fitz/fitz_wrap.o
  fitz/fitz_wrap.c:2767:10: fatal error: fitz.h: No such file or directory
   #include <fitz.h>
            ^~~~~~~~
  compilation terminated.
  error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for pymupdf
  Running setup.py clean for pymupdf
Failed to build pymupdf
Installing collected packages: pymupdf
    Running setup.py install for pymupdf ... error
    ERROR: Command errored out with exit status 1:
     command: /home/pi/dev/guidegf/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/setup.py'"'"'; __file__='"'"'/tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-wgla2cn2/install-record.txt --single-version-externally-managed --compile --install-headers /home/pi/dev/guidegf/include/site/python3.7/pymupdf
         cwd: /tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/
    Complete output (20 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-armv7l-3.7
    creating build/lib.linux-armv7l-3.7/fitz
    copying fitz/__init__.py -> build/lib.linux-armv7l-3.7/fitz
    copying fitz/fitz.py -> build/lib.linux-armv7l-3.7/fitz
    copying fitz/utils.py -> build/lib.linux-armv7l-3.7/fitz
    copying fitz/__main__.py -> build/lib.linux-armv7l-3.7/fitz
    running build_ext
    building 'fitz._fitz' extension
    creating build/temp.linux-armv7l-3.7
    creating build/temp.linux-armv7l-3.7/fitz
    arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -I/usr/include/mupdf -I/usr/local/include/mupdf -Imupdf/thirdparty/freetype/include -I/home/pi/dev/guidegf/include -I/usr/include/python3.7m -c fitz/fitz_wrap.c -o build/temp.linux-armv7l-3.7/fitz/fitz_wrap.o
    fitz/fitz_wrap.c:2767:10: fatal error: fitz.h: No such file or directory
     #include <fitz.h>
              ^~~~~~~~
    compilation terminated.
    error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/pi/dev/guidegf/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/setup.py'"'"'; __file__='"'"'/tmp/pip-install-uxmi137a/pymupdf_8e4bd29fef5b46e9be53199a6ba101da/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-wgla2cn2/install-record.txt --single-version-externally-managed --compile --install-headers /home/pi/dev/guidegf/include/site/python3.7/pymupdf Check the logs for full command output.

Really appreciate the help.

**Comment by JorjMcKie** [comment]
These errors indicate, that MuPDF has not been previously installed.
Either this - or the PyMuPDF `setup.py` has not been adjusted to point to the correct MuPDF installation libraries.

**Comment by leonardoee** [comment, marked as answer]
Hi Jorj,

I'm having trouble to install MuPDF. Below is the error I have when I run `make prefix=/usr/local install`:



pi@pi:/gcc_all/mupdf $ make prefix=/usr/local install
    CC build/debug/pdf/pdf-pkcs7.o
source/pdf/pdf-pkcs7.c: In function ‘bsegs_read’:
source/pdf/pdf-pkcs7.c:30:39: error: dereferencing pointer to incomplete type ‘BIO’ {aka ‘struct bio_st’}
  BIO_SEGS_CTX *ctx = (BIO_SEGS_CTX *)b->ptr;
                                       ^~
source/pdf/pdf-pkcs7.c: At top level:
source/pdf/pdf-pkcs7.c:107:1: error: variable ‘methods_bsegs’ has initializer but incomplete type
 static BIO_METHOD methods_bsegs =
 ^~~~~~
source/pdf/pdf-pkcs7.c:109:2: warning: excess elements in struct initializer
  0,"segment reader",
  ^
source/pdf/pdf-pkcs7.c:109:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:109:4: warning: excess elements in struct initializer
  0,"segment reader",
    ^~~~~~~~~~~~~~~~
source/pdf/pdf-pkcs7.c:109:4: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:110:2: warning: excess elements in struct initializer
  NULL,
  ^~~~
source/pdf/pdf-pkcs7.c:110:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:111:2: warning: excess elements in struct initializer
  bsegs_read,
  ^~~~~~~~~~
source/pdf/pdf-pkcs7.c:111:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:112:2: warning: excess elements in struct initializer
  NULL,
  ^~~~
source/pdf/pdf-pkcs7.c:112:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:113:2: warning: excess elements in struct initializer
  NULL,
  ^~~~
source/pdf/pdf-pkcs7.c:113:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:114:2: warning: excess elements in struct initializer
  bsegs_ctrl,
  ^~~~~~~~~~
source/pdf/pdf-pkcs7.c:114:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:115:2: warning: excess elements in struct initializer
  bsegs_new,
  ^~~~~~~~~
source/pdf/pdf-pkcs7.c:115:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:116:2: warning: excess elements in struct initializer
  bsegs_free,
  ^~~~~~~~~~
source/pdf/pdf-pkcs7.c:116:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:117:2: warning: excess elements in struct initializer
  bsegs_callback_ctrl,
  ^~~~~~~~~~~~~~~~~~~
source/pdf/pdf-pkcs7.c:117:2: note: (near initialization for ‘methods_bsegs’)
source/pdf/pdf-pkcs7.c:135:17: error: field ‘x509_ctx’ has incomplete type
  X509_STORE_CTX x509_ctx;
                 ^~~~~~~~
source/pdf/pdf-pkcs7.c: In function ‘verify_callback’:
source/pdf/pdf-pkcs7.c:159:13: error: dereferencing pointer to incomplete type ‘X509_STORE_CTX’ {aka ‘struct x509_store_ctx_st’}
  switch (ctx->error)
             ^~
source/pdf/pdf-pkcs7.c: In function ‘pk7_verify’:
source/pdf/pdf-pkcs7.c:208:40: error: dereferencing pointer to incomplete type ‘X509_STORE’ {aka ‘struct x509_store_st’}
  X509_VERIFY_PARAM_set_flags(cert_store->param, X509_V_FLAG_CB_ISSUER_CHECK);
                                        ^~
source/pdf/pdf-pkcs7.c: At top level:
source/pdf/pdf-pkcs7.c:107:19: error: storage size of ‘methods_bsegs’ isn’t known
 static BIO_METHOD methods_bsegs =
                   ^~~~~~~~~~~~~
source/pdf/pdf-pkcs7.c: In function ‘bsegs_ctrl’:
source/pdf/pdf-pkcs7.c:66:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
source/pdf/pdf-pkcs7.c: In function ‘bsegs_callback_ctrl’:
source/pdf/pdf-pkcs7.c:105:1: warning: control reaches end of non-void function [-Wreturn-type]
 }
 ^
make: [Makefile:136: build/debug/pdf/pdf-pkcs7.o] Error 1


**Would installing 64bits Raspian solve these issues?**

**Reply by JorjMcKie** [reply]
Maybe, but I am not sure. The error does not look like that.
Also, you should probably ask that question at MuPDF channels.
Generally, you must of course compile MuPDF with the same C compiler - which you did not as far as I can see.
As I said, I am experimenting with ARM version wheels, and there now is one [here](https://github.com/pymupdf/PyMuPDF-wheels/blob/linux-extra/PyMuPDF-1.18.10-cp39-cp39-manylinux2014_aarch64.whl) for ARM 64bit and Python 3.9. Whether or not that would run on a 64bit Raspberry remains to be seen, but is worth trying and would save you all this compilation huzzles.

**Reply by n3m4x** [reply]
Hey @JorjMcKie,

I have a similar problem. I'm trying to install pymupdf (with pip3 install pymupdf) on a Pi 3 B with an aarch64 architecture. How do you recommend proceeding in my situation?

**Reply by JorjMcKie** [reply]
@n3m4x - sorry I forgot to generate the ARM wheels. Give me an hour or so. I'll let you know when they are on PyPI.

**Reply by JorjMcKie** [reply]
The wheels are on PyPI now, naming convention `PyMuPDF-1.18.17-cp39-cp39-manylinux_2_17_aarch64.manylinux2014_aarch64.whl`.

**Reply by n3m4x** [reply]
@JorjMcKie - it works now, thank you very much!!!

**Reply by topshed** [reply]
Are these ARM wheels still available, please?

**Reply by JorjMcKie** [reply]
@topshed - they will come back officially with the new version 1.19.5 in the next few days.
You can already now download your wheel as a pre-version from [here](https://github.com/pymupdf/PyMuPDF-wheels/tree/master/wheelhouse).

**Reply by viertel97** [reply]
@JorjMcKie Have the aarch64 wheels for the version 1.19.5 already been published? I am receiving the same error as OP with my Raspberry Pi 4 (Linux raspberrypi 5.10.17-v8+ #1403 SMP PREEMPT Mon Feb 22 11:37:54 GMT 2021 aarch64 GNU/Linux) when i run "pip3 install pymupdf". I installed "mupdf" and "mupdf-tools" via "sudo apt-get install" before, but it still does not work.

**Reply by JorjMcKie** [reply]
Yes they have. Please have a look at [PyPI](https://pypi.org/project/PyMuPDF/#files) to confirm.

**Reply by viertel97** [reply]
@JorjMcKie Thank you!
I can confirm, that files are there, but with every "pip3 install *_aarch64.whl" (with or without sudo) file i receive the error "[filename] is not a supported wheel on this platform". The aarch64 packes should be right, shouldnt they?

**Reply by JorjMcKie** [reply]
1. I would **_never_** just execute `pip3`, but instead always use the correct Python and then do `python -m pip ...`.
2. You must make sure that you have the **_current_** pip. So best always execute `python -m pip install -U pip` first.
3. Have you confirmed that the aarch64 wheel is there with that same Python version?

**Reply by viertel97** [reply]
I did all the steps you mentioned. The newest version is installed and i used the cp37 wheel which is for the python 3.7 version, am i correct?
If i try it, this is the output:

```
pi@raspberrypi:~ $ uname -a
Linux raspberrypi 5.10.17-v8+ #1403 SMP PREEMPT Mon Feb 22 11:37:54 GMT 2021 aarch64 GNU/Linux

pi@raspberrypi:~ $ python3 -m pip --version
pip 22.0.3 from /home/pi/.local/lib/python3.7/site-packages/pip (python 3.7)

pi@raspberrypi:~ $ python3 -m pip install PyMuPDF-1.19.5-cp37-cp37m-manylinux_2_17_aarch64.manylinux2014_aarch64.whl
Defaulting to user installation because normal site-packages is not writeable
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
ERROR: PyMuPDF-1.19.5-cp37-cp37m-manylinux_2_17_aarch64.manylinux2014_aarch64.whl is not a supported wheel on this platform.
```
A `sudo` before the `python3` removes the `Defaulting to user installation because normal site-packages is not writeable`.

**Reply by JorjMcKie** [reply]
Hm, indeed - everything looks as it should. Out of ideas.
Last resort: build from sources. Try to follow the instructions in the `installation` folder for Linux / Ubuntu, if you want to do that.

**Reply by viertel97** [reply]
That worked for me! It showed some errors, but after that i could install PyMuPDF via `python3 -m pip install PyMuPDF`.
Thank you very much for your support!

**Comment by topshed** [comment]
Thank you. Would it be possible to include 32bit versions for Raspberry Pi OS?

**Reply by JorjMcKie** [reply]
> Would it be possible to include 32bit versions for Raspberry Pi OS?

No, I am afraid. Does not seem to be supported by [cibuildwheel](https://github.com/pypa/cibuildwheel).
I recommend you generate from sources locally.
